### ISSUE TITLE
Using Joint Fitness NMS

### ISSUE THREAD

**Issue by Oussamab21** [lachlants/denet#6]
Hi , Thank you for sharing this great work 
I want to use the function that implements Joint Fitness NMS in other implementation what should I do ?

Thank you

**Comment by lachlants** [comment]
Hello, 

I haven't pushed the Joint Fitness NMS code to this repo yet... was waiting for CVPR, however I've uploaded the file which has pretty much all the changes here: [denet_detect.py](https://gist.github.com/lachlants/574c5cd9ffebda7efe8dd76675197bf0) . It's a bit messy but should give you a pretty good idea how I'm constructing the distributions, etc. 

When I get time I'll merge with this repo and upload some models.

**Comment by Oussamab21** [comment]
Hello,
Thank you I appreciate it.

**Comment by twmht** [comment]
@lachlants 

When are you planing to upload the code?

thanks

**Comment by lachlants** [comment]
@Oussamab21 @twmht  Hey, sorry for the wait. I've uploaded the code and 2x models for testing. Let me know if you have any issues.

**Comment by twmht** [comment]
Very appreciate:)

**Comment by ryanlouc** [comment]
Hi,

@lachlants I read your Fitness NMS paper, but I can't figure out what the training objective of the Fitness NMS yet. Can you explain where is the label of Fitness NMS while training?

**Comment by lachlants** [comment]
Hello, 

The Joint Fitness method adds another axis to the class distribution generated for each region of interest (RoI) e.g. instead of estimating Pr(class | RoI) we estimate Pr(class, fitness | RoI). In this implementation, I merge the class and fitness into a single axis and set the label via `label = C*5 + int(min(10*(IoU-0.5), 4))` where IoU is the intersection over union of the RoI  and the closest groundtruth bounding box, and C is it's class. Note that any RoI with IoU < 0.5 is set to the null class so label >= 0. 

You can see this occuring here:
https://github.com/lachlants/denet/blob/3337c54168b925d1f9d221e5df0201503b2e61c1/denet/layer/denet_detect.py#L180-L183

I use the standard softmax + negative log likelihood cost between the target and predicted Pr(class, fitness | RoI) distributions.

**Comment by ryanlouc** [comment]
> Hello,
> 
> The Joint Fitness method adds another axis to the class distribution generated for each region of interest (RoI) e.g. instead of estimating Pr(class | RoI) we estimate Pr(class, fitness | RoI). In this implementation, I merge the class and fitness into a single axis and set the label via `label = C*5 + int(min(10*(IoU-0.5), 4))` where IoU is the intersection over union of the RoI and the closest groundtruth bounding box, and C is it's class. Note that any RoI with IoU < 0.5 is set to the null class so label >= 0.
> 
> You can see this occuring here:
> [denet/denet/layer/denet_detect.py](https://github.com/lachlants/denet/blob/3337c54168b925d1f9d221e5df0201503b2e61c1/denet/layer/denet_detect.py#L180-L183)
> 
> Lines 180 to 183 in [3337c54](/lachlants/denet/commit/3337c54168b925d1f9d221e5df0201503b2e61c1)
> 
>  if self.use_jointfit: 
>      f = max(0, min(int(self.fitness_num*sample_f), self.fitness_num-1)) 
>      det_pr[b, sample_cls*self.fitness_num + f, sample_j, sample_i] = 1.0 
>      det_pr[b, self.null_class, sample_j, sample_i] = 0.0 
> I use the standard softmax + negative log likelihood cost between the target and predicted Pr(class, fitness | RoI) distributions.

Oh, I got it. Thanks :)
